### PR TITLE
Fix type issue in using bucketing with Trainer

### LINKED ISSUE #18003: Optional type of lengths causes slow speed in LengthGroupedSampler
### System Info

- `transformers` version: 4.20.1
- Platform: macOS-12.4-arm64-arm-64bit
- Python version: 3.8.13
- Huggingface_hub version: 0.8.1
- PyTorch version (GPU?): 1.12.0 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: no

but this bug doesn't depend on environment.

### Who can help?

@sgugger 

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
import time
import torch
import random

megabatches = [[random.randint(1, 512) for _ in range(51200)] for _ in range(2)]

# test lengths_list_int
lengths_torch_tensor = torch.tensor([random.randint(1, 512) for _ in range(102400)])

start = time.time()
megabatches_ = [list(sorted(megabatch, key=lambda i: lengths_torch_tensor[i], reverse=True)) for megabatch in megabatches]
end = time.time()
print(end-start)

# test lengths_list_int
lengths_torch_tensor = torch.tensor([random.randint(1, 512) for _ in range(102400)])
lengths_list_int = lengths_torch_tensor.tolist()

start = time.time()
megabatches_ = [list(sorted(megabatch, key=lambda i: lengths_list_int[i], reverse=True)) for megabatch in megabatches]
end = time.time()
print(end - start)
```
```bash
1.0904111862182617
0.013269901275634766
```

### Expected behavior

When using `group_by_length`, `length_column_name` in `TrainingArguments`, `get_length_grouped_indices` function in `LengthGroupedSampler` is very slow if `Dataset[length_cloumn_name]` is `torch.Tensor(List[int])` (e.g. `torch.Tensor([200,100,..])`). So I think that `lengths: Optional[List[int]]` -> `lengths: List[int]` in `__init__` method of `LengthGroupedSampler` or warning message is printed and type casting to `List[int]`.

https://github.com/huggingface/transformers/blob/49c8c67fb815a277405f84dea4a66353e19fb347/src/transformers/trainer_pt_utils.py#L532-L569

https://github.com/huggingface/transformers/blob/49c8c67fb815a277405f84dea4a66353e19fb347/src/transformers/trainer_pt_utils.py#L520

### PR DESCRIPTION
**Pull request by seopbo**
# What does this PR do?

- Fix type issues in `LengthGrouperSampler`, `DistributedLengthGroupedSampler`

Fixes #18003

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger

### CODE REVIEW

File: src/transformers/trainer_pt_utils.py [hunk 1]
```diff
@@ -614,6 +622,14 @@ def __init__(
                     f"'{model_input_name}' key."
                 )
             lengths = [len(feature[model_input_name]) for feature in dataset]
+        else:
+            if isinstance(lengths, torch.Tensor):
+                logger.warning(
```
Comment (sgugger) [inline src/transformers/trainer_pt_utils.py:627]:
```suggestion
                logger.info(
```

**Review by sgugger — APPROVED** [review]
Thanks for your PR! As this will happen naturally when a Dataset has been fully converted to torch tensors, I think we should just issue an info instead of a warning. Otherwise LGTM!

File: src/transformers/trainer_pt_utils.py (hunk 1 above)
Comment (seopbo) [reply src/transformers/trainer_pt_utils.py:627]: I agree this. thanks @sgugger

File: src/transformers/trainer_pt_utils.py [hunk 2]
```diff
@@ -558,6 +558,14 @@ def __init__(
                     f"'{model_input_name}' key."
                 )
             lengths = [len(feature[model_input_name]) for feature in dataset]
+        else:
+            if isinstance(lengths, torch.Tensor):
+                logger.info(
+                    "If lengths is a torch.Tensor, LengthGroupedSampler will be slow. Converting lengths to"
+                    " List[int]..."
+                )
+                lengths = lengths.tolist()
```
Comment (sgugger) [inline src/transformers/trainer_pt_utils.py:567]:
```suggestion
        elif isinstance(lengths, torch.Tensor):
            logger.info(
                "If lengths is a torch.Tensor, LengthGroupedSampler will be slow. Converting lengths to"
                " List[int]..."
            )
            lengths = lengths.tolist()
```

**Review by sgugger — APPROVED** [review]
Oh and sorry I didn't catch it at the first review, but we can group the else if in an elif.

**Pull request merged**